### PR TITLE
End to end test/profile screen

### LINKED ISSUE #174: End-to-end test 6: the user login, and select a song, and add it to his top track playlist

### PR DESCRIPTION
**Pull request by mbouaissi**
This test the fact that the user login, arrive at the main screen, then goes in the profile screen, then goes in the select a song screen, select a song, and goes back to the main screen. 

closes #174

### CODE REVIEW

File: app/src/androidTest/java/ch/epfl/cs311/wanderwave/viewmodel/ProfileViewModelTest.kt [hunk 1]
```diff
@@ -90,20 +92,38 @@ class ProfileViewModelTest {
...
-    assertFalse("Song list should not be empty after adding a track", songLists.isEmpty())
 
-    songLists = viewModel.songLists.value
-    // Check if the track was added correctly
-    val songsInList = songLists.find { it.name == "TOP SONGS" }?.tracks ?: emptyList()
-    Log.d("Temp", songsInList.toString())
+    // Optionally check additional conditions after ensuring Flow had time to collect
+    assertFalse(
+        "Song list should not be empty after adding a track", viewModel.songLists.value.isEmpty())
+    Log.d("fwefewfew", viewModel.songLists.value.first().tracks.toString())
```
Comment (KlaKalma) [inline app/src/androidTest/java/ch/epfl/cs311/wanderwave/viewmodel/ProfileViewModelTest.kt:122]: Use copilot for this, it's very usefull to have the logs having the right title and not something random like this. Can you change it ?

File: app/src/main/java/ch/epfl/cs311/wanderwave/ui/screens/SelectSongScreen.kt [hunk 2]
```diff
@@ -50,12 +45,13 @@ fun SelectSongScreen(navActions: NavigationActions, viewModel: ProfileViewModel)
 
   var displayedList by remember { mutableStateOf(mainList) }
 
-  LaunchedEffect(Unit) { viewModel.retrieveAndAddSubsection() }
+  LaunchedEffect(Unit) { viewModel.retrieveAndAddSubsection(CoroutineScope(Dispatchers.IO)) }
```
Comment (KlaKalma) [inline app/src/main/java/ch/epfl/cs311/wanderwave/ui/screens/SelectSongScreen.kt:48]: why not the view model scope ? Why do you need this ?

File: app/src/main/java/ch/epfl/cs311/wanderwave/viewmodel/ProfileViewModel.kt [hunk 3]
```diff
@@ -115,8 +114,8 @@ constructor(
    * @since 2.0
    * @last update 2.0
    */
-  fun retrieveTracks() {
-    CoroutineScope(Dispatchers.IO).launch {
+  fun retrieveTracks(scope: CoroutineScope) {
+    scope.launch {
```
Comment (KlaKalma) [inline app/src/main/java/ch/epfl/cs311/wanderwave/viewmodel/ProfileViewModel.kt:118]: Is it important that you send the scope ? Seems a bit weird that you control this from the other screen. This is just a question, there probably is a reason, but just to know

**Review by KlaKalma — CHANGES_REQUESTED** [review]
Nice code, thanks for being the first one to tackle an end-to-end test and have everything working so we can get inspiration. There seem to be some code that is not related to the test. Probably from another of your branches. I put some comments on things I think are good to change. Thanks for your work

File: app/src/androidTest/java/ch/epfl/cs311/wanderwave/endToEnd/LoginAndAddSong.kt [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+
+  @get:Rule val mockkRule = MockKRule(this)
+
+  @RelaxedMockK private lateinit var mockNavigationActions: NavigationActions
+
+  @RelaxedMockK private lateinit var mockViewModel: SpotifyConnectScreenViewModel
+
+  lateinit var viewModel: ProfileViewModel
+
+  @RelaxedMockK private lateinit var profileRepositoryImpl: ProfileRepositoryImpl
```
[the remark this answers is not in the document]
Comment (mbouaissi) [reply app/src/androidTest/java/ch/epfl/cs311/wanderwave/endToEnd/LoginAndAddSong.kt:57]: idk what we are going to use, so i let it like this for now

[the remark this answers is not in the document]
Comment (KlaKalma) [reply app/src/androidTest/java/ch/epfl/cs311/wanderwave/endToEnd/LoginAndAddSong.kt:57]: It's fine I'll delete it when we merge

**Pull request merged**